### PR TITLE
[STRATCONN-144] Ability to map top level fields to Adobe 

### PR DESCRIPTION
**Pull request by briemcnally**
**What does this PR do?**
JIRA: https://segment.atlassian.net/browse/STRATCONN-144

Add the ability to map a list of top level properties (currently only messageId and anonymousId) to Adobe Analytics props, eVars, or context data variables. 

**Are there breaking changes in this PR?**
No. 

**Any background context you want to provide?**
This is an ask from both Intuit and Fox to improve their ability to join data from their Warehouses with Adobe Reports. 

**Is there parity with the server-side/android/iOS integration components (if applicable)?**
No. This will need to be implemented on Server Side, iOS and Android to maintain parity. Will create StratConn tickets as a followup action item. 

**Does this require a new integration setting? If so, please explain how the new setting works**
No. 

**Links to helpful docs and other external resources**
Escalated urgency in response to Fox<>Adobe Discrepancy. Details in this doc: https://paper.dropbox.com/doc/Fox-Adobe-Discrepancies-between-Adobe-and-Redshift--AzCREu_W0y_PjJtcip_it0SiAQ-orfqhwMtwHM1TL6ji2RWx

### CODE REVIEW

File: integrations/adobe-analytics/lib/index.js [hunk 1]
```diff
@@ -666,7 +663,7 @@ function clearKeys(keys) {
  * @param {Object} options
  */
 
-function extractProperties(props, options) {
+function extractProperties(facade, options, propType) {
```
[the remark this answers is not in the document]
Comment (briemcnally) [reply integrations/adobe-analytics/lib/index.js:666]: Yes will update

File: integrations/adobe-analytics/lib/index.js [hunk 2]
```diff
@@ -676,13 +673,39 @@ function extractProperties(props, options) {
...
+    // On track and page calls the propType here will be empty. On video HB calls the propTyp will
+    // be 'mergedPropContext'. On those events we use  the extracted properties from this fxn to explicitly
+    // map link vars on `window.s` object.
+    if (topLevelProperties.includes(mappedKey) && propType !== 'context') {
+      value = facade.proxy(mappedKey);
+    }
+
+    // When we are checking the context object for segment property mappings (updateContextData)
+    // we only want set top level fields to Context Data Variables.
+    var contextValueKeys = Object.keys(options.contextValues);
```
[the remark this answers is not in the document]
Comment (briemcnally) [reply integrations/adobe-analytics/lib/index.js:701]: So `extractProperties` is used many ways. 
1. For `track()` and `page()` calls we pass this fxn the `properties` object and use the result to update link vars on the `window.s` object
2. For HB events we access both the `properties` and `context` object and use the result to update custom content metadata on video events. 
3. When updating context data. We update context data on `track()`, `page()`, and all e-commerce events. These `contetxtValues` from our destination settings  are `segmentContextField: adobeContextDataField`. For these we have always looked in just the context object. There is a bug in our implementation that unfortunately now must be maintained [that you can see here](https://github.com/segmentio/analytics.js-integrations/pull/465/files/2c34180bc58909e687d4566ec2cc4d9ab00c6526#diff-e7435b76f43719e786ff94a43acc1f6eR569-L574). When we go to update context data fields we actually also look at eVars, props, hVars, lVars etc. These usually always return undefined because their `mappedKey` is defined in the properties object not the context object. 

Basically if the toplevelProperty is `anonymousId` or `messageId` and we are NOT setting context data values we can set `value = facade.proxy(mappedKey);`. Otherwise we only want to set that if the a customer defines they want that top-level mapping to update Context Data Fields by checking their `contextValues` settings.  Otherwise we could incorrectly set `contextData.eVar5` when we update context data.

This code definitely needs a larger refactor IMO, but this will solve the immediate need from Fox and Intuit.

[the remark this answers is not in the document]
Comment (gpsamson) [reply integrations/adobe-analytics/lib/index.js:704]: Got it. Thanks for the in-depth write-up. I totally overlooked the bug we need to maintain. 

> This code definitely needs a larger refactor IMO, but this will solve the immediate need from Fox and Intuit.

Agreed. I think this is all 🟢 to merge.
